tests/suites/deploy/deploy_bundles.sh

```diff
@@ -54,22 +54,63 @@ run_deploy_cmr_bundle() {
 	destroy_model "other"
 }
 
-run_deploy_exported_bundle() {
+# run_deploy_exported_charmstore_bundle_with_fixed_revisions is to test which checks
+# how juju deploys charmstore bundles with fixed revisions
+run_deploy_exported_charmstore_bundle_with_fixed_revisions() {
 	echo
 
-	file="${TEST_DIR}/test-export-bundles-deploy.log"
+	file="${TEST_DIR}/test-export-bundles-deploy-with-fixed-revisions.log"
 
-	ensure "test-export-bundles-deploy" "${file}"
+	ensure "test-export-bundles-deploy-with-fixed-revisions" "${file}"
 
 	bundle=./tests/suites/deploy/bundles/telegraf_bundle.yaml
 	juju deploy ${bundle}
 
 	# no need to wait for the bundle to finish deploying to
 	# check the export.
 	juju export-bundle --filename "${TEST_DIR}/exported-bundle.yaml"
-	diff ${bundle} "${TEST_DIR}/exported-bundle.yaml"
+	diff -u ${bundle} "${TEST_DIR}/exported-bundle.yaml"
 
-	destroy_model "test-export-bundles-deploy"
+	destroy_model "test-export-bundles-deploy-with-fixed-revisions"
+}
+
+# run_deploy_exported_charmhub_bundle_with_float_revisions is to test which checks
+# how juju deploys charmhub bundles with undefined versions in bundle yaml
+run_deploy_exported_charmhub_bundle_with_float_revisions() {
+  echo
+
+  file="${TEST_DIR}/test-export-bundles-deploy-with-float-revisions.log"
+
+  ensure "test-export-bundles-deploy-with-float-revisions" "${file}"
+  bundle=./tests/suites/deploy/bundles/telegraf_bundle_without_revisions.yaml
+  bundle_with_fake_revisions=./tests/suites/deploy/bundles/telegraf_bundle_with_fake_revisions.yaml
+  juju deploy ${bundle}
+
+  if ! which "yq" >/dev/null 2>&1; then
+    sudo snap install yq --classic --channel latest/stable
+  fi
+
+  echo "Get current stable revisions for charms in telegraf_bundle_without_revisions.yaml"
+  influxdb_rev=$(juju info influxdb --format json | jq -r '."channel-map"."latest/stable".revision')
+  telegraf_rev=$(juju info telegraf --format json | jq -r '."channel-map"."latest/stable".revision')
+  ubuntu_rev=$(juju info ubuntu --format json | jq -r '."channel-map"."latest/stable".revision')
```

Comment:
This seems correct today. Looking at the data, though, I don't see that this will stay correct.
Specifically if I look at:
```
    "latest/stable": {
      "released-at": "2021-10-25T14:56:46.348903+00:00",
      "track": "latest",
      "risk": "stable",
      "revision": 24,
      "size": 7316781,
      "version": "24",
      "architectures": [
        "amd64",
        "arm64",
        "armhf",
        "i386",
        "ppc64el",
        "s390x"
      ],
      "series": [
        "bionic",
        "focal",
        "trusty",
        "xenial"
      ]
```

The *design* of Charmhub channels is that you would have a *different* revision for amd64 and arm64, and a different revision for bionic vs focal.
This may be *Juju* modeling it wrong, but what we really need is something akin to:
```
"revisions': {
  latest/stable/amd64/ubuntu/20.04: 24
  latest/stable/amd64/ubuntu/22.04: 25
  latest/stable/arm64/ubuntu/20.04: 26
}
```

I'm actually quite concerned seeing this data, because it is very different from how we really need to be modeling it.
(You're ok for implementing the test in this way, but its a sign we have something fundamentally wrong)

Reply:
I got your concern and agree that it sounds fundamental. 

Can we, in the current juju CLI implementation, get a charm revision in the more precise and right way?

Reply:
And what do you think about using flags `--series` and `--arch` in `juju info`?

The series we know from the bundle description and the arch from test Jenkins description. By specifying info about arch and series, we will get more accurate info about revision numbers.